### ISSUE TITLE
Logo strips

### ISSUE THREAD

**Issue by lyubomir-popov** [canonical-web-and-design/vanilla-framework#2847]
@marcushas @waynecrosby @bartaz @spencerbygraves 

*** This issue is a placeholder for the discussion we had in the Working Group today, because there wasn't an issue created for it. Just to serve as a place to track progress on this.

## Pattern to amend

Two and a half years ago we came up with a process to improve logo strips on our sites. The idea was to move away from the inline images pattern, and replace it with grid-based markup that doesn't require special classes. While this was happening for a short period, recently we see many instances of reverting to the old way of doing logo strips, and also many hybrid versions (or strips that were improved, then partially regressed).

Legacy inline images:

<img width="1418" alt="Screenshot 2020-02-24 at 17 14 11" src="https://user-images.githubusercontent.com/2741678/75175814-efb3c380-572a-11ea-8fff-eb11f32a53f3.png">

New grid based approach:
<img width="1196" alt="Screenshot 2020-02-24 at 17 38 45" src="https://user-images.githubusercontent.com/2741678/75176782-b54b2600-572c-11ea-9fde-a8112f9d1ddd.png">

<img width="1460" alt="Screenshot 2020-02-24 at 17 21 18" src="https://user-images.githubusercontent.com/2741678/75175847-fd694900-572a-11ea-80c6-a30d167620b8.png">

Hybrid with regressions:

<img width="1361" alt="Screenshot 2020-02-24 at 17 11 53" src="https://user-images.githubusercontent.com/2741678/75175741-d27ef500-572a-11ea-8953-ec65b72ccd82.png">

To recap, the issues with the legacy inline images pattern are:

Problems with the inline images pattern:
- disregards the grid, e.g. 5 logos on a 12 column grid
- makes no attempt to control the scale of logos 
- is agnostic of any text inside the logos, and rules for turning arbitrary logos into a coherent sequence

Problems with the new grid based system:
- developers find it not simple enough ( not sure if that means difficult to understand, or time consuming to type )

Process problems:
Design is often not involved when a product manager or marketing requests a logo be changed. This means the logo doesn't get resized / repositioned within its bounding box so that it is "coherent" with its companion logos within the strip, which leads to the hybrid regressions strip example above.

TBD:
- Understand what makes the new system cumbersome and propose a solution

**Comment by bartaz** [comment]
How is 'new system' supposed to work exactly?
As far as I see from snapcraft example it's not possible without [adding custom styles](https://github.com/canonical-web-and-design/snapcraft.io/blob/master/static/sass/_snapcraft_custom-cols.scss)

Without custom styles logos are not centered in second line (as on maas.io):

<img width="1217" alt="Screenshot 2020-02-25 at 09 00 27" src="https://user-images.githubusercontent.com/83575/75226839-57aaee00-57ad-11ea-83d6-283026fcaed7.png">

I don't see Vanilla having any documented examples right now on how to use grid to make these logo lists. This may be the main cause of why it's hard to adopt it. It's much easier to modify existing legacy ones.

To the contrary, we have a documented pattern for inline images: https://docs.vanillaframework.io/patterns/inline-images

<img width="1186" alt="Screenshot 2020-02-25 at 09 06 47" src="https://user-images.githubusercontent.com/83575/75227288-2d0d6500-57ae-11ea-8175-1ebce5d0ea59.png">

There is no mention of this not being a recommended pattern to use.

If we don't want anyone to use the 'legacy' styles, we should deprecate the old pattern and provide documentation explaining how to build these lists with the grid.

If there are any additional classes needed to center the logos when they wrap into two lines, Vanilla should provide them.

And we need to make sure we have examples that cover different cases (different number of logos, wrapped into two lines, or not, approaches to style for smaller screens).

**Comment by lyubomir-popov** [comment]
> How is 'new system' supposed to work exactly?

Using the grid col-start-* classes: 
https://docs.vanillaframework.io/patterns/grid#empty-columns

> As far as I see from snapcraft example it's not possible without [adding custom styles](https://github.com/canonical-web-and-design/snapcraft.io/blob/master/static/sass/_snapcraft_custom-cols.scss)

Snapcraft's case is a special one - [lengthy discussion here](https://github.com/canonical-web-and-design/vanilla-framework/issues/1974)): In addition to the 3 breakpoints in vanilla, s.io required a fourth change to avoid logos getting too small. Since this request has only happened once, it was implemented with custom classes. If this need arises elsewhere, then maybe we should provide a fourth breakpoint in vanilla, which would allow us to avoid any custom classes.

> Without custom styles logos are not centered in second line (as on maas.io):

TLDR: Easy to fix with the above vanilla classes, see image below.
Some additional context: This is because of the "process problems" I mentioned: updates that omit design and go from product manager directly to developer.  We somehow went from 24 logos in 2 strips that alternate on reload to just 10 logos. Had that gone through design, I would have suggested the following addition (just 3 classes from vanilla, no custom css, seen at the highlighted row and two rows further down):
![image](https://user-images.githubusercontent.com/2741678/75236666-03583c00-57b6-11ea-8e28-24719b062187.png)

https://docs.vanillaframework.io/patterns/grid#empty-columns

> I don't see Vanilla having any documented examples right now on how to use grid to make these logo lists. This may be the main cause of why it's hard to adopt it. It's much easier to modify existing legacy ones.

Fair point, we should update the documentation. Historically, my focus has been on creating a reference implementation (the maas.io being the most recent version of that), but this is no substitute for documentation. I can do that as part of this issue, or you're welcome to help if you have time for it.

**Comment by bartaz** [comment]
@lyubomir-popov Thanks for the comment.

Even that I was aware of `col-start-` classes being in Vanilla I haven't thought that they can be used for that. Possibly because our current grid examples are very abstract.

We need a better, more practical examples of grid usage. Logos list seems like a perfect fit for that.
I think we should move mass.io example into Vanilla examples (making the second line centered with `col-start-` classes). 

And in addition to that we should deprecate `p-inline-images` pattern. Unless we think there are some cases when it should be used instead of grid? In such case we should clearly document it.

**Comment by lyubomir-popov** [comment]
I agree.  

As to deprecating inline-images, are we sure the logo strips the only thing they are used for?I have a vague memory of seeing an inline list of social icons or something similar using that pattern. If that's no longer the case, I would vote to deprecate them.

In terms of logo examples, I have 2 thoughts:
- The number of logos in a strip can change at anytime (usually because a company requests their logo be removed for legal reasons). This means a 12 logo strip can become an 11, 10, 9, 8, 7, 6, 5, 4 or 3 logo strip at any time. Examples should show the preferred layout at each breakpoint. This means showing how to handle any number of logos from 12 down to 3, on 4, 6 and 12 columns.

- Among developers here there's a reluctance to add multiple classes on a single element for "aesthetic reasons". Doing it on 12+ consecutive elements is even harder to swallow, and I think this may have something to do with the reluctance to use this system.

While I disagree with this guidance, there's a way to reduce the number of classes required per strip to just 1 (instead of say 36 (12logos, 3 classes on each). For example, we could bundle cases using a shorthand notation:

`.p-logo-strip-2-3-6`: display 2 logos per row on the mobile, 3 on tablet, 6 on desktop
`.p-logo-strip--2-2c-4c`: Indicate alignment with a letter after the number, e.g. 2 per row on the 4 col grid, 2 centered on the 6col grid, 4 centered on the desktop 12col grid
... and so on

**Comment by bartaz** [comment]
> Among developers here there's a reluctance to add multiple classes on a single element for "aesthetic reasons". Doing it on 12+ consecutive elements is even harder to swallow, and I think this may have something to do with the reluctance to use this system.

I wouldn't worry about this. I would rather have multiple classes that are easier to read and understand rather than some magic `.p-logo-strip--2-2c-4c` that we would need to generate and support.
If you give developers class name with multiple numbers in it, you can bet they will assume they can change these numbers however they want, so it would be expected to have all the combinations covered.

> Examples should show the preferred layout at each breakpoint.

Not sure if we need to provide example for every single case, but for sure the guidance what to do in case of odd/even number of logos, and extremes (3 vs 12 or more) would be needed.

**Comment by lyubomir-popov** [comment]
@bartaz Either way is fine by me, whatever people prefer. 

FWIW, I can tell you from personal experience the shorthand notation can speed up your work 10 fold, but it does have a half-hour learning curve which may not be to everyone's taste.

**Comment by bartaz** [comment]
As for the usage of `p-inline-images`:
https://github.com/search?l=HTML&q=p-inline-images+user%3Acanonical-web-and-design&type=Code

It's used quite a lot, mostly ubuntu.com. But by clicking around I haven't found one place that wouldn't be a logo list that probably should go into grid.

**Comment by lyubomir-popov** [comment]
Sorry to keep banging on, but here's a very real example from Maas I weant to mention. 

Having say 36 classes on 12 consecutive elements is likely to tempt a future PR called "clean up markup" that removes everything but the single class for the largest breakpoint (this has actually happened in MAAS). This kind of pr is also very likely to be enthusiastically approved without full QA, as it is perceived as a performance optimisation ( again, something that actually happened in MAAS ).

A system that has one unfamiliar class is much less likely to become a victim to this kind of "optimisation".

**Comment by bartaz** [comment]
That's a fair point, but I don't think it's the problem we can solve on Vanilla level (although we can try to reduce number of classes needed).

PRs that change markup should get Design +1. It seems that that one didn't. Or it wasn't QAed at all on smaller screens.

**Comment by anthonydillon** [comment]
I will keep this short as I have argued this so many times and leave it to you two to decide from now but these are my concerns:
- Many classes means a large learning curve
- Many classes can lead to mixing them in strange ways that result in broken layouts. Vanilla can no longer rely on mark up being in a predictable way.
- This tightly couples the icon layout with the grid pattern
- This means a designer has to bespokely create and state the grid classes for each logo list which will increase site development time
- How should we handle randomly generated partner logo lists?

As long these concerns have been considered, I will see what you decide.

**Comment by bartaz** [comment]
> This tightly couples the icon layout with the grid pattern

As far as I understand, from design point of view, this is what we want - we want logos on the grid. That's the whole issue this pattern is trying to solve. So it will directly (by using grid classes) or indirectly (by using grid placeholders, or something similar internally) depend on grid.

As for the implementation part and number of classes:

Why current `p-inline-images` is so widely used? Because it's simple. You add a pattern, create a list of images and it doesn't matter how many images there are. You can freely add or remove images from the list without affecting any other logos.

When we tie elements to the grid it stops being that simple. Elements on the list can't have all the same style and class. The styles applied to them depend on how many elements are on the list and which element on the list it is.

So, if we implement it using existing grid classes, it means that each logo on the list will need to have multiple classes applied to it and these classes may need to change when new elements are added or removed from logos, or order will change. This is going to be hard to understand and will easily lead to errors. The benefit - CSS is kept to minimum (these grid classes already exist) and there is freedom to layout these logos is specific ways required by the design.

If we want to keep the simplicity similar to `p-inline-images` while holding images on the grid it will complicate the CSS part if Vanilla, possibly quite a lot. Because of styles of every image on the list depend on the number of images on the list and which image it is it will be very hard (if possible) to implement it without any additional classes added to the elements.

I guess we would need to experiment on what it feasible and find a compromise between ease of development, amount of CSS required, and control the design has on the placement of logos.

**Comment by bartaz** [comment]
@lyubomir-popov try to experiment with your proposal

**Comment by lyubomir-popov** [comment]
Matching the grid column width and gap width in a flex layout gets us close enough, no classes needed other than the one on the component itself. Will clean this up and submit a per tomorrow:

![image](https://user-images.githubusercontent.com/2741678/112358914-a97fb980-8cc8-11eb-8ad2-c525cb0a03d9.png)
![image](https://user-images.githubusercontent.com/2741678/112358973-b7cdd580-8cc8-11eb-95d2-473ce093b6f8.png)
![image](https://user-images.githubusercontent.com/2741678/112359073-cfa55980-8cc8-11eb-8c28-b72d1aab3ce8.png)
